File: day12/src/main.rs
use std::{collections::HashSet, env, fs};

const DIRECTIONS: [(i32, i32); 4] = [(0, -1), (0, 1), (-1, 0), (1, 0)];

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        print!(
            "Usage: {executable_name} <filename.txt>\n",
            executable_name = args.get(0).unwrap()
        );
        return;
    }
    let filename = args.get(1).unwrap();
    let contents = fs::read_to_string(filename).unwrap();

    let grid: Vec<Vec<char>> = contents
        .lines()
        .map(|line| line.chars().collect())
        .collect();
    let width = grid.get(0).unwrap().len();
    let height = grid.len();

    println!("Part 1: {}", part1(&grid, width as i32, height as i32));
    println!("Part 2: {}", part2(&grid));
}

fn part1(grid: &Vec<Vec<char>>, width: i32, height: i32) -> i32 {
    let claimed: HashSet<(usize, usize)> = HashSet::new();

    for (y, line) in grid.iter().enumerate() {
        for (x, ch) in line.iter().enumerate() {
            if claimed.contains(&(x, y)) {
                continue;
            }

            let mut perimeter = 0;
            let mut area = 0;
            let mut queue: Vec<(usize, usize)> = Vec::new();

            for delta in DIRECTIONS {
                let np = (x as i32 + delta.0, y as i32 + delta.1);
                if !check_point_inbounds(np, width, height) {
                    continue;
                }
                if &grid[np.1 as usize][np.0 as usize] != ch {
                    perimeter += 1;
                }
            }
        }
    }
    0
}

fn check_point_inbounds(point: (i32, i32), width: i32, height: i32) -> bool {
    return 0 <= point.0 && point.0 < width && 0 <= point.1 && point.1 < height;
}

fn part2(input: &Vec<Vec<char>>) -> i32 {
    0
}

File: day12/src/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class Main {

    public record Point(int x, int y) {
    }

    public static final Point[] DIRECTIONS = new Point[]{
        new Point(-1, 0), new Point(1, 0),
        new Point(0, -1), new Point(0, 1)
    };

    public static void main(final String[] args) throws IOException {
        if (args.length != 1) {
            System.err.println("Usage: java Day12 <filename.txt>");
            return;
        }

        final String fileName = args[0];
        final String contents = Files.readString(Path.of(fileName));
        final List<List<Character>> grid = parseInput(contents);
        final int part1Answer = part1(grid);
        System.out.printf("Part 1: %d\n", part1Answer);
    }

    public static int part1(final List<List<Character>> grid) {
        final Set<Point> claimed = new HashSet<>();
        final int width = grid.get(0).size();
        final int height = grid.size();
        int total = 0;

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                final Point source = new Point(x, y);
                if (claimed.contains(source)) continue;

                int perimeter = 0;
                int area = 0;
                final Queue<Point> queue = new LinkedList<>();
                final Set<Point> visited = new HashSet<>();
                queue.add(source);
                final char oldChar = grid.get(source.y).get(source.x);

                while (!queue.isEmpty()) {
                    final Point currPoint = queue.poll();

                    if (!isPointInbound(currPoint, width, height)) {
                        perimeter++;
                        continue;
                    }

                    final char newChar = grid.get(currPoint.y).get(currPoint.x);
                    if (visited.contains(currPoint)) continue;

                    if (oldChar != newChar) {
                        perimeter++;
                        continue;
                    }
                    area++;
                    claimed.add(currPoint);
                    visited.add(currPoint);

                    final List<Point> neighbors = getNeighbors(currPoint);
                    queue.addAll(neighbors);
                }
                total += perimeter * area;
            }
        }
        return total;
    }

    public static List<Point> getNeighbors(final Point p) {
        final List<Point> neighbors = new ArrayList<>();
        for (final Point dir : DIRECTIONS) {
            neighbors.add(new Point(
                p.x + dir.x, p.y + dir.y
            ));
        }
        return neighbors;
    }

    public static boolean isPointInbound(final Point p, final int width, final int height) {
        return 0 <= p.x && p.x < width && 0 <= p.y && p.y < height;
    }

    public static List<List<Character>> parseInput(final String input) {
        return input
            .lines()
            .map(line -> line
                .chars()
                .mapToObj(c -> (char) c)
                .toList()
            )
            .toList();
    }
}